Basic functional test

dcermak opened the pull request:
This is my current state of the openQA tests for the kiwi staging area.
This fixes #2, #5, #6

What it does:
- iso images are booted, login is tested and either the vm is turned off, if this is an installation iso or rebooted and reloged in, if it is a live iso
- disk images are booted, user is logged in, rebooted, user logs in again and then the VM is turned off
- the created hard drive of an installation iso is saved and goes through the same test as any other kiwi disk image

what currently does not work:
- ~~login on the openstack qcow2 image is broken, the default credentials do not work~~ (fixed by @schaefi, the image had its root console deactivated)
- all the raw disk images appear not to work, for some reason openQA cannot create a working disk image from these
- kiwi-test-image-disk-legacy.x86_64.iso and kiwi-test-image-custom-partitions.x86_64.iso hang during the installation and never reach a login prompt
- the hackery that I added to select the correct entry in the boot menu is not really that great and does only appear to work by accident

What still needs to be done:
- [x] documentation what is going on and how to configure openqa (external settings are necessary)
- ~~[ ] add some actually meaningful testing, currently we just test whether this image boots~~
- [x] test other distributions as well
- [x] reconsider whether we should include the build id of the images in the schedule command: calling them by their static link means that openQA does not download newer versions if the old asset stays around (this should be done)

schaefi commented:
@dcermak I added the "do not merge" label since you are still working on it. Please drop the label when you are done. Thanks

okurz commented on tests/boot.pm:
```diff
@@ -19,17 +19,37 @@ use strict;
-    assert_script_run(
-        "systemd-analyze blame | head -n 15 | tee -a /dev/$serialdev"
-    );
+    assert_screen("kiwi_bootloader");
+
+    # for some live images the default selected boot entry is the "Boot from Hard Disk" entry
+    # we don't want that, so we need to select the correct entry that says "Install $NAME"
+    # to achieve that, we just try to spam first up and then down until we get a needle match
+    if (check_screen("kiwi_bootloader_boot_from_hdd") && defined(get_var("PUBLISH_HDD_1"))) {
```
These both `assert_screen … check_screen` calls can likely be combined
with a multi-tag `assert_screen` with `match_has_tag` instead of `check_screen` with non-zero timeout to prevent introducing any timing dependant behaviour, to save test execution time as well as state more explicitly from the testers point of view what are the expected alternatives. A common simple example that I supply should explain it:

```perl
assert_screen([qw(yast2_console-finished yast2_missing_package)]);
if (match_has_tag('yast2_missing_package')) {
    send_key 'alt-o';  # confirm package installation
    assert_screen 'yast2_console-finished';
}
```

so let me try to suggest what you could here, e.g. like this:

```perl
assert_screen [qw(kiwi_bootloader kiwi_bootloader_boot_from_hdd)];
if match_has_tag 'kiwi_bootloader_boot_from_hdd' && defined(get_var("PUBLISH_HDD_1")) {
…
```

okurz commented on tests/login.pm:
```diff
@@ -0,0 +1,43 @@
+sub run {
+    assert_screen("login_prompt");
+    type_string("root");
+    send_key("ret", wait_screen_change => 1);
```
While using `wait_screen_change` and `wait_still_screen` sound simpler in some cases they often either waste time or introduce timing behaviour dependant problems. `wait_screen_change` means that this line finishes as soon as any screen change is seen, which might before the system is ready to accept the next line where you type the password.

It is often better to be explicit and have a proper needle check in the middle:

```perl
assert_screen 'login_prompt';
type_string 'root';
send_key 'ret';
assert_screen 'password_prompt';
type_string 'linux';
send_key 'ret';
assert_screen 'textmode_logged_in';
```

I see that you have added code further down to handle "grub stuff on the screen". Likely your needle for 'textmode_logged_in' is too big and hence too restrictive as it might fail to be detected due to the "grub stuff". If that is the case think like a human. What do you *really* look for on the screen to detect that the "textmode is logged in"? Likely it is just something like the prompt sign and some free space to the right of it, e.g. `#     ` and nothing else :) This is what is included in the corresponding needle within os-autoinst-distri-opensuse

okurz commented on tests/reboot.pm:
```diff
@@ -0,0 +1,30 @@
+use testapi;
+
+sub run {
+    assert_screen("textmode_logged_in");
+    assert_script_run("reboot");
+
+    assert_screen("kiwi_bootloader");
+
+    assert_screen("login_prompt");
```
just a matter of taste. I would write no parentheses for such simple code and use single ticks to prevent accidental variable expansion where not explicitly desired, e.g.:

```perl
    assert_screen 'textmode_logged_in';
    assert_script_run 'reboot';
    assert_screen 'kiwi_bootloader';
    assert_screen 'login_prompt';
```

dcermak replied:
The single tick backticks are a good idea, thank you!

But unless @schaefi insists on leaving out the parenthesis, I'd really like to keep them (I just don't like this shell like parameter passing convention in perl).

schaefi replied:
I agree with you on the parenthesis. As I come from python and formerly from C and also wrote perl in the C style syntax I'm used to parenthesis for function parameter lists. Nevertheless I have no super strong opinion about it if we do it consistently. So parenthesis everywhere or nowhere but no mix please :)

I also checked on perlcritic in --brutal mode and it does not complain if you use parenthesis or not. On the other side it does complain if you use builtin methods like e.g print with parenthesis. This means if you want to handle this topic in a consistent way and perlcritic --brutal clean code you are forced to prevent parenthesis unless a custom .perlcritic file configures the desired behavior.

In the end this is personal taste and I only want consistency if that makes sense to you too

okurz commented on tests/reboot.pm:
```diff
@@ -0,0 +1,30 @@
+    assert_screen('textmode_logged_in');
+    assert_script_run('reboot');
+
+    assert_screen('kiwi_bootloader');
```
If this is only temporarily shown, e.g. due to an automatic timeout I recommend to not have that check at all due to the possibility that the test system might "miss" the screen and then fail when the SUT already continues booting. This happened in other cases hence I am sharing that problem from the past. A good hint how to design openQA tests in this regard: Think about how a manual tester would type in "reboot" and then goes away to get a cup of coffee. After return the test would see the 'login_prompt' only, not the intermediate state. If you ask the overly stupid tester to "check for kiwi_bootloader" that tester would say the test failed as the login prompt is shown instead of the expected "kiwi_bootloader" screen :)

By the way, openQA automatically records screenshots if the screen changes more than a defined threshold so it's likely that the bootloader screen would show up as intermediate screenshot anyway regardless of an explicit check or not.

dcermak replied:
Hm, this makes sense, but then again, not checking for the bootloader could also result in the situation that you just press ctrl+d and never reboot. The test would then still pass, as it only checks that the login screen is present and newer for the bootloader. But not sure whether it really makes sense to worry about that.

okurz replied:
If you want to be explicit then disable the grub timeout before reboot, trigger reboot, wait for grub menu, and explicitly call `send_key 'ret'`.

Basically I see three options:

1. leave it as is
2. Add the explicit grub menu check with disabled timeout and explicit boot confirmation action needed
3. reboot, wait for login and check afterwards that a reboot actually happened, e.g. in logs or uptime
4. just remove the check and not know if the system actually rebooted

I am just providing the options and of course I am in no position to "block" your PR for this repo here :) But based on my experience I just advise against 1. and invest the effort to go for 2., 3. or 4. because I am convinced that sooner or later 1. will cause more pain and time wasted to understand what is going on than what one would consider worth the effort :)

okurz commented on settings.py:
```diff
@@ -0,0 +1,220 @@
+            ):
+                # FIXME: we can drop this once
+                # https://github.com/os-autoinst/openQA-python-client/pull/17
+                # is merged
```
done now https://github.com/os-autoinst/openQA-python-client/pull/17

dcermak replied:
A new release would be terrific for that ;-)

okurz replied:
Created new release in github with https://github.com/os-autoinst/openQA-python-client/releases/tag/4.1.2 if that's what you mean? Or do you mean the package in OBS? Don't know how that is commonly done

dcermak replied:
No, I meant the package on PyPI, that's what poetry pulls down.

okurz replied:
done by AdamWill: https://pypi.org/project/openqa-client/4.1.2/

okurz commented on tests/shutdown.pm:
```diff
@@ -0,0 +1,29 @@
+    assert_screen('textmode_logged_in');
+    # the SUT can shutdown faster than the reply reaching the worker via the
+    # serial line, which causes a failure, albeit the SUT has been turned off
+    script_run('shutdown -hP now', timeout => 0);
```
```suggestion
    enter_cmd 'shutdown -hP now';
```

okurz commented on tests/reboot.pm:
```diff
@@ -0,0 +1,31 @@
+    # faster in restarting the SUT than the serial device reporting success (and
+    # then assert_script_run throws an error although it actually did the right
+    # thing)
+    script_run('reboot', timeout => 0);
```
```suggestion
    enter_cmd 'reboot';
```

okurz commented on tests/boot.pm:
```diff
@@ -19,17 +19,54 @@ use strict;
+    # bootloader is at the wrong entry.
+    # We do this in a second check_screen, to ensure that the matched needle
+    # will not be "just" 'kiwi_bootloader'.
+    check_screen([qw(kiwi_bootloader_boot_from_hdd trust_uefi_certificates)]);
```
`check_screen` should only be used if you check the return code, otherwise you can end up with weird followup errors and confusing "red boxes" in test details. If I read the code correctly the next best screen you look for is `login_prompt` so you could do

```suggestion
    # wait 10 minutes for the unattended installation and 2 for a regular boot
    # (booting can take a while for appliances based on MicroOS as ignition
    # takes its time)
    my $timeout = defined(get_var('PUBLISH_HDD_1')) ? 600 : 120;
    assert_screen([qw(kiwi_bootloader_boot_from_hdd trust_uefi_certificates login_prompt)], $timeout);
    …
    assert_screen('login_prompt', $timeout) unless match_has_tag 'login_prompt';
```

okurz commented on pyproject.toml:
```diff
@@ -0,0 +1,26 @@
+lxml-stubs = ">=0.1.1"
+
+[tool.black]
+line-length = 79
```
out of curiosity, why not stay with black's default of 88 characters?

dcermak replied:
Because PEP8 says 79 :wink:

okurz commented on main.pm:
```diff
@@ -20,6 +20,16 @@ use autotest;
 
 testapi::set_distribution(distribution->new);
-autotest::loadtest "tests/boot.pm";
+autotest::loadtest('tests/boot.pm');
+autotest::loadtest('tests/login.pm');
+
+if (defined(get_var('PUBLISH_HDD_1'))) {
+    autotest::loadtest('tests/shutdown.pm');
+} else {
+    autotest::loadtest('tests/reboot.pm');
+    autotest::loadtest('tests/boot.pm');
+    autotest::loadtest('tests/login.pm');
+    autotest::loadtest('tests/shutdown.pm');
+}
```
I suggest a simple loading wrapper

```suggestion
sub loadtest { autotest::loadtest('tests/' . $_[0]) }

loadtest('boot');
loadtest('login');

if (defined(get_var('PUBLISH_HDD_1'))) {
    loadtest('shutdown');
} else {
    loadtest('reboot');
    loadtest('boot');
    loadtest('login');
    loadtest('shutdown');
}
```

the extension `.pm` is automatically appended

okurz commented:
@dcermak I am happy to review your changes again but I would appreciate if you can first describe the changes since your last push for easier review.

okurz commented on tests/shutdown.pm:
```diff
@@ -0,0 +1,29 @@
+# Copyright (C) 2020 SUSE LLC
```
s/2020/2020-2021/

okurz commented on tests/boot.pm:
```diff
@@ -19,17 +19,54 @@ use strict;
+    if (match_has_tag('trust_uefi_certificates')) {
+        die 'trust UEFI certificates screen present, but UEFI is not used' unless defined(get_var('UEFI'));
+
+        if (match_has_tag('trust_uefi_certificates_no')) {
+            send_key('down');
+        }
```
my style would be

```suggestion
        send_key 'down' if match_has_tag 'trust_uefi_certificates_no';
```

dcermak commented:
> @dcermak I am happy to review your changes again but I would appreciate if you can first describe the changes since your last push for easier review.

@okurz I have been mostly working on the actual python launcher and on the test configuration of the tests. The launcher has been now refactored two or three times and at this point I am finally sorta-kinda happy with it: now it schedules tests based on the actual packages in the staging projects and in contrast to using regexes on the published repos (which was a pretty dumb idea from my side).

The tests themselves have actually not changed a lot:
- the yaml job template is nearly the same, only that I removed the `+ISO_1=""` and `+ISO_1_URL=""` in the chained tests of installed disk images. I originally thought that since openqa inherits the `ISO_1` variable, it would boot from the iso instead of the created hdd, but that was not the case, so the removal is fine.
- the boot test now handles even more stuff: we have LUKS based disk images that require a passphrase to be entered and the MicroOS disk image runs some extensive post installation that needs requires longer wait times

There's a few things where I'd like to get your input as a experienced test writer & maintainer:
- I have started recording soft failures for grub not showing up in UEFI mode which happens fairly frequently. I am however not really sure whether that is such a great idea as there is now a **ton** of soft failures.
- How would you recommend to check for a successful reboot? I am mostly relying on the bootloader showing up again, but that does not always happen (see the above point).
- There is currently a few broken disk images (see #8). Assuming that these are really broken: how would you recommend to handle these failures?

dcermak commented:
@schaefi @okurz @davidcassany @Conan-Kudo How should we proceed with this further? I am testing all of this at the moment only on my local machine, should we maybe go forward and create a staging/development project on o3 and see how much breaks there?

okurz commented:
@dcermak sure you can use o3. For this I recommend to not use any of the "old ways" of putting test distributions into o3 but use http://open.qa/docs/#_triggering_tests_based_on_an_any_remote_git_refspec_or_open_github_pull_request

asmorodskyi commented:
haven't yet look into code , but 2 general comments : 
1. This PR looks too big monster which would generate too many discussions about too many different topics . Maybe it will be more effective to split it into several smaller ? 
2. Not sure if it is really good idea to mix needles and code repos into one ...

schaefi commented:
yes this PR has turned more into a conversation. I think Dan had not intended to merge it after the many side tracks. I assume this one will close and @dcermak wants to merge in smaller chunks what he has done on his local testing setup

@dcermak please correct if I'm wrong

dcermak commented:
> yes this PR has turned more into a conversation. I think Dan had not intended to merge it after the many side tracks. I assume this one will close and @dcermak wants to merge in smaller chunks what he has done on his local testing setup
> 
> @dcermak please correct if I'm wrong

Honestly, I would prefer to merge it like this once I manage to get it to run on o3. I can clean up the git history a bit further, but I don't really see how all this can be really reduced down into "manageable chunks".

schaefi commented:
ok, in this case let's cleanup the git history and go for it as an initial kick-off. As this will be the start I think it's ok. For further changes we should however, work in readable, smaller change sets.

@dcermak feel free to merge as you see fit and we can go forward from that state.

Thoughts ?

schaefi commented:
@dcermak btw, welcome back :) hope you had a good vacation ?

okurz commented:
@dcermak as over the past months you have updated this PR but it does not look like you update to have it merged I will unsubscribe from the PR to not get repeated notifications. Feel free to ping me again as soon as you updated the PR for actual review.

dcermak commented:
@okurz we have decided during this week's kiwi community meeting that we are going to drop this PR and just merge the whole thing via #10 as soon as the tests succeed on my local box so that others can collaborate and help get them to run on o3.
